src/hmem_cuda_gdrcopy.c

```diff
@@ -292,6 +293,45 @@ void cuda_gdrcopy_from_dev(uint64_t handle, void *hostptr,
 			  GDRCOPY_FROM_DEVICE);
 }
 
+static ssize_t cuda_gdrcopy_iov_buf(uint64_t handle,
```

Comment:
I am not sure about this. Can you extend `ofi_copy_to/from_hmem_iov` to cover gdrcopy?

Reply:
That is possible but I want to get consensus from the community.

To extend `ofi_copy_to/from_hmem_iov` for gdrcopy, we need to:
1. Define the behavior in the cuda->cuda scenario - the caller could pass cuda buffers for both src and dest, but that is not allowed for gdrcopy. This is currently a segfault, and it is not trivial to make it graceful - we need a mechanism to detect that both src and dest are cuda buffers.
1. Pass `hmem_data` alongside `device` to e.g. `ofi_copy_to_hmem_iov`. That's very doable and I'm ok with that. But it requires changes in multiple place. Just want to make sure the community is equally happy before I do it.

So, my current plan is to lift gdrcopy from `ofi_copy_to/from_hmem_iov` path, meaning:
1. `ofi_copy_to/from_hmem_iov` will only use `cudaMemcpy`, which supports cuda->cuda.
2. The caller can opt to call `cuda_gdrcopy_to/from` only when they know what they are doing. This essentially makes gdrcopy an optional copy utility.

Reply:
ofi_copy_to/from_hmem_iov() only copy between device and host memory.  Device-to-device or host-to-host copies are not supported.  The 'buf' parameter into those calls is always a host buffer.

Passing in hmem_data is more complicated.

Regardless, I think the intent is for the caller to request gdrcopy, which is better handled explicitly, versus calling an abstraction that hides its use.  That is, either hide gdrcopy and the decision to use it lower in the stack, or expose it and let the decision to use it be controlled higher up.  The current approach seems to follow the latter, so, I agree with @wenduwan that it makes sense to keep this path separate.

I don't like that we duplicate the implementation of ofi_copy_hmem_iov_buf() below, but it's honestly only about 3 lines of code that end up duplicated.

Reply:
> ofi_copy_to/from_hmem_iov() only copy between device and host memory. Device-to-device or host-to-host copies are not supported. The 'buf' parameter into those calls is always a host buffer.

Oh that was opposite to my observation. I see SHM is using those routines on ipc path: https://github.com/ofiwg/libfabric/blob/main/prov/shm/src/smr_progress.c#L598-L607

But if we make the intention explicit, i.e. only D2H/H2D are supported, then I see gdrcopy can be nicely hidden inside it.

But still we need to pass around `hmem_data`. Trying to pick @shefty 's brain on:
- Do we want to enforce `ofi_copy_to/from_hmem_iov() only copy between device and host memory`?
- Mind talking more about `Passing in hmem_data is more complicated`? Coding wise it sounds simple, but verbose. What other complexity do you have in mind for this change?